### PR TITLE
Implement @woocommerce/price-format dependency

### LINKED ISSUE #28: Add @woocommerce/price-format dependency
Currently, the extension is getting the price format directly from the `cart` prop. To prevent reinventing the wheel, the price should be formatted using the `@woocommerce/price-format` dependency.

See also https://github.com/woocommerce/woocommerce-free-shipping-progress-bar-block/pull/27#discussion_r734393504 and p1634901323077500/1634868391.069600-slack-C8X6Q7XQU.

* [x] Add helper functions `getCurrencyFromPriceResponse()` and `formatPrice()`.
* [ ] Create mock file to load default values in the editor.

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #28

### Note

The remaining price should be formatted using the @woocommerce/price-format dependency, which is included in the WooCommerce Blocks plugin.

#### Before all test cases

1. Check out this PR.
2. Create a test page and add the Cart block.
3. Insert the Free Shipping Progress Bar block into the Cart block and save the page.
4. Set the free shipping value within the  Free Shipping Progress Bar block to 1200.
5. Head over to the frontend and add products to your cart.
6. Ensure the cart total is higher than 1000 (to test the thousand separator) and below 1200 (to see the remaining price).

#### Test 0 decimals currency

1. Go to `WP Admin » WooCommerce » Settings » Currency options` and use the following settings:
    * Currency: `Indonesian rupiah (Rp)`
    * Currency position: `Left`
    * Thousand separator: `.`
    * Decimal separator: `,`
    * Number of decimals: `0`
2. Go to the cart and verify that the price is formatted as expected.

#### Test 2 decimals currency

1. Go to `WP Admin » WooCommerce » Settings » Currency options` and use the following settings:
    * Currency: `USD $`
    * Currency position: `Left + space`
    * Thousand separator: `,`
    * Decimal separator: `.`
    * Number of decimals: `2`
2. Go to the cart and verify that the price is formatted as expected.

#### Test 3 decimals currency

1. Go to `WP Admin » WooCommerce » Settings » Currency options` and use the following settings:
    * Currency: `Libyan Dinar`
    * Currency position: `Right + space`
    * Thousand separator: `.`
    * Decimal separator: `,`
    * Number of decimals: `3`
2. Go to the cart and verify that the price is formatted as expected.

### CODE REVIEW

File: src/components/progress-label/index.js [hunk 1]
```diff
@@ -1,13 +1,32 @@
...
 	freeShippingFrom,
 	labelInsufficientTotals,
 	labelSufficientTotals,
+	cart,
 } ) => {
-	const remaining = Number( freeShippingFrom - currentTotal ).toFixed( 2 );
+	const currentTotal = getCurrentTotal( cart );
+	const minorUnit = getMinorUnit( cart );
+	const remaining = getRemaining( freeShippingFrom, currentTotal, minorUnit );
+	const currency = getCurrencyFormat( cart );
```
Comment (senadir) [inline src/components/progress-label/index.js]: This seems like an early abstraction that shouldn't happen yet until you actually need it, further more, some of functions are useless if you check for cart here.

File: src/util.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+export function getCurrentTotal( cart ) {
+	if (
+		undefined !== cart &&
+		undefined !== cart.cartTotals.total_price &&
+		undefined !== cart.cartTotals.currency_minor_unit
+	) {
+		const total_price = cart.cartTotals.total_price;
+		const minor_unit = cart.cartTotals.currency_minor_unit;
+
+		return total_price / Math.pow( 10, minor_unit );
```
Comment (senadir) [inline src/util.js:17]: Why you formatting here? shouldn't this be `formatPrice` job?

File: src/util.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ * Get the current totals.
+ *
+ * @param {Object} cart The cart object.
+ * @returns {Number} The totals of the cart object, if available, else the totals of the cart progress bar preview.
+ */
+export function getCurrentTotal( cart ) {
+	if (
+		undefined !== cart &&
+		undefined !== cart.cartTotals.total_price &&
+		undefined !== cart.cartTotals.currency_minor_unit
```
Comment (senadir) [inline src/util.js]: When cart is defined, those are always defined, they're the in the default response.

File: src/util.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ */
+export function getMinorUnit( cart ) {
+	if (
+		undefined !== cart &&
+		undefined !== cart.cartTotals.currency_minor_unit
+	) {
+		return cart.cartTotals.currency_minor_unit;
+	}
+	return '';
+}
```
Comment (senadir) [inline src/util.js]: This is also not needed.

File: src/util.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+ * @param {Number} currentTotal The current total value.
+ * @param {Number} minorUnit The minor unit value.
+ * @returns {Number} The difference between freeShippingFrom and currentTotal.
+ */
+export function getRemaining( freeShippingFrom, currentTotal, minorUnit ) {
+	const remaining = freeShippingFrom - currentTotal;
+	const power = Math.pow( 10, minorUnit );
+	const result = remaining.toFixed( minorUnit ) * power;
+
+	return result;
```
Comment (senadir) [inline src/util.js:46]: What's the point of using power here?

File: src/util.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+/**
+ * Get the currency format.
+ *
+ * @param {Object} cart The cart object.
+ * @returns {Object} The currency format object.
+ */
+export function getCurrencyFormat( cart ) {
```
Comment (senadir) [inline src/util.js]: There seems to be a lot of redundant logic here, you can get currency by using `getCurrencyFromPriceResponse( cart.cartTotals )`

**Review by senadir — CHANGES_REQUESTED** [review]
I'm finding it hard to follow what's happening, I left some comments about (what I believe to be) redundant code and eager abstraction.

To my mind, this is how it should be done:
1. Convert the string provided from `"50"` to `5000`. This is, to my knowledge, not something that `@woocommerce/format-price provide, but is something we're doing in Filter By Product block.
2. Do the math.
3. Format result back.

This should all happen directly on the component, no need for extra utils that are one time use, depending on how or easy 1 is, it might get its own utility function to reduce clutter, the rest can stay.

File: src/components/progress-label/index.js (hunk 1 above)
Comment (nielslange) [reply src/components/progress-label/index.js]: I'm not sure what you mean with the early abstraction. Could you elaborate on that? And could you tell me which functions you think are useless here? I created them to extract the cart total, the minor units and the currency format from the cart object, if a cart object is available.

File: src/util.js (hunk 2 above)
Comment (nielslange) [reply src/util.js:17]: The `formatPrice()` function expects the price in minor units based on the selected number of decimals. 

- If a merchant selected 2 decimals, a price of USD 50 equals a minor unit price of 5000. 
- If a merchant selected 3 decimals, a price of USD 50 equals a minor unit price of 50000. 

With `return total_price / Math.pow( 10, minor_unit );` I ensure that I'm calculating the correct minor unit price.

File: src/util.js (hunk 3 above)
Comment (nielslange) [reply src/util.js]: Thanks for pointing this out. I've removed these duplicated checks.

File: src/util.js (hunk 4 above)
Comment (nielslange) [reply src/util.js]: I also removed these duplicated checks.

File: src/util.js (hunk 5 above)
Comment (nielslange) [reply src/util.js:46]: Similar as before, I need the remaining value as minor unit price. I'm using pow to calculate the minor unit price based on the number of decimals.

File: src/util.js (hunk 6 above)
Comment (nielslange) [reply src/util.js]: Good catch. I removed my function `getCurrencyFormat()` and I'm using `getCurrencyFromPriceResponse()` instead.

**Comment by nielslange** [comment]
> I'm finding it hard to follow what's happening, I left some comments about (what I believe to be) redundant code and eager abstraction.
> 
> To my mind, this is how it should be done:
> 1. Convert the string provided from `"50"` to `5000`. This is, to my knowledge, not something that `@woocommerce/format-price provide, but is something we're doing in Filter By Product block.
> 2. Do the math.
> 3. Format result back.

That's pretty much what I'm doing here. As for converting `"50"` to `5000`, this would only be correct when the merchant selected 2 decimals to display. For 3 decimals, the minor unit price should be 50000 and for 0 decimals, the minor unit price should remain 50.

> This should all happen directly on the component, no need for extra utils that are one time use, depending on how or easy 1 is, it might get its own utility function to reduce clutter, the rest can stay.

I moved the helper functions into `util.js`, as I'm calling the function `getCurrentTotal()` both in `src/components/progress-bar/index.js` and in `src/components/progress-label/index.js`. In addition, it keeps the individual components leaner which, in my opinion, improves both readability and maintainability.

As I'm using `getMinorUnit()` and `getRemaining()` only in `src/components/progress-label/index.js`, I could technically move these two functions directly into the corresponding component. I decided against that, as I believe it makes more sense to have all related functions in one file instead of spreading them across the project. 

I'm looking forward to hearing your thoughts on my feedback regarding your review and the additional notes I added before.

File: src/util.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Get the minor unit.
+ *
+ * @param {Object} cart The cart object.
+ * @returns {Number} The totals of the cart object, if available, else the totals of the cart progress bar preview.
+ */
+export function getMinorUnit( cart ) {
+	if ( undefined !== cart ) {
+		return cart.cartTotals.currency_minor_unit;
```
Comment (mikejolley) [inline src/util.js]: nit, but to also protect against invalid cart response, you could write this function like so:

```
return cart?.cartTotals?.currency_minor_unit || '';
```

This will not throw an error if the cart is undefined, and will always default to an empty string should the minor unit be missing.

File: src/util.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	}
+	return '';
```
Comment (mikejolley) [inline src/util.js]: Why do we return a string? The function says number, so perhaps return `0`?

File: src/util.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ * @returns {Number} The totals of the cart object, if available, else the totals of the cart progress bar preview.
+ */
+export function getCurrentTotal( cart ) {
+	if ( undefined !== cart ) {
+		const total_price = cart.cartTotals.total_price;
+		const minor_unit = cart.cartTotals.currency_minor_unit;
+
+		return total_price / Math.pow( 10, minor_unit );
+	}
+	return 15;
```
Comment (mikejolley) [inline src/util.js]: I think this is pretty confusing as a return value. Could the preview data not be hardcoded using the `useEditorContext` hook in your block? Mixing real and preview data is dangerous IMO.

**Review by mikejolley — APPROVED** [review]
Changes look good, and Nadirs feedback was resolved. I made some small suggestions but nothing blocking. Test coverage also good.

File: src/util.js (hunk 7 above)
Comment (nielslange) [reply src/util.js]: Thanks for this great suggestion! I'll definitely use this pattern.

File: src/util.js (hunk 8 above)
Comment (nielslange) [reply src/util.js]: Good question why I'm return a string. Does not make any sense here, and I changed it already.

File: src/util.js (hunk 9 above)
Comment (nielslange) [reply src/util.js]: I was returning a default value to render the progress bar component correctly in the editor, as the editor does not have the cart object. That said, I see how this default value can be confusing, and I'm now returning the default value of the `block.json` file.

**Comment by nielslange** [comment]
I applied all changes and fixed the integration tests, which broke due to adding `@woocommerce/price-format` in this change. Do you want to run another review, @mikejolley?
